File: content/code/go-gettext/po2json.go
package po2json

import "regexp"
import "io/ioutil"
import "encoding/json"

type msgIdStrPairs map[string]string
type localesMsg map[string]msgIdStrPairs

const pattern = `msgid "(.+)"\nmsgstr "(.+)"`

func getPOPath(locale, domain, localedir string) string {
	return localedir + "/" + locale + "/LC_MESSAGES/" + domain + ".po"
}

func extractFromPOFile(popath string) msgIdStrPairs {
	buf, err := ioutil.ReadFile(popath)
	if err != nil {
		panic(err)
	}

	re := regexp.MustCompile(pattern)
	matches := re.FindAllStringSubmatch(string(buf), -1)

	pairs := msgIdStrPairs{}
	for _, array := range matches {
		pairs[array[1]] = array[2]
	}
	return pairs
}

func PO2JSON(locales []string, domain, localedir string) string {
	// create PO-like json data for i18n
	obj := localesMsg{}
	for _, locale := range locales {
		// English is default language
		if locale == "en_US" {
			continue
		}

		obj[locale] = extractFromPOFile(getPOPath(locale, domain, localedir))
	}

	b, err := json.Marshal(obj)
	if err != nil {
		panic(err)
	}

	return string(b)
}

File: content/code/python-i18n-gettext/po2json.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

import re
import json

def getPOPath(locale, domain, localeDir):
  return localeDir + "/" + locale + "/LC_MESSAGES/" + domain + ".po"

def extractFromPOFile(poPath):
  with open(poPath, 'r') as f:
    tuples = re.findall(r'msgid "(.+)"\nmsgstr "(.+)"', f.read())
  return tuples

def PO2JSON(locales, domain, localeDir):
  # create PO-like json data for i18n
  obj = {}
  for locale in locales:
    # English is default language
    if locale == "en_US": continue

    obj[locale] = {}
    tuples = extractFromPOFile( getPOPath(locale, domain, localeDir) )
    for tuple in tuples:
      obj[locale][tuple[0].decode('utf-8')] = tuple[1].decode('utf-8')
      #obj[locale][tuple[0]] = tuple[1]

  return json.dumps(obj)

if __name__ == '__main__':
  locales = ["zh_TW", "vi_VN"]
  domain = "messages"
  localeDir = "locale"
  print(PO2JSON(locales, domain, localeDir))
